### PR TITLE
feat(Core/Spells): improve SpellCheckRange(), spell cast when dismounting, falling or levitating spells

### PR DESCRIPTION
**Pull request by Helias**
imported Maczuga/SunwellCore@5d9826c

## CHANGES PROPOSED:
-  enable item use and spell cast when dismounting.
- improve SpellCheckRange()
- spell cast when dismounting
- falling or levitating spells
- Make certain spells (e.g. Mind Control) properly flag the caster for PvP

## ISSUES ADDRESSED:
- related #1141

## HOW TO TEST THE CHANGES:
I am not sure about the steps to tests these changes, but I can figure out that the followings step can help:
- use the levitate spell of Priest and the mage one for levitating
- try to fall from a high place
- try to fall dismounting from a flying mount
- try to fall dismounting when you are jumping from a high place and dismounting and use spells
- cast spells while are you in mount
- try to use spells (Melle and Ranged, so use Warrior, Hunter and Mage maybe) and use from very far (also if the client show that you can't) or from near NPCs
- use Mind Control in duel against other player and on NPC, use it once in battleground, check if everything seems ok

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp [hunk 1]
```diff
@@ -3210,10 +3210,9 @@ void AuraEffect::HandleAuraModIncreaseFlightSpeed(AuraApplication const* aurApp,
         // do not remove unit flag if there are more than this auraEffect of that kind on unit on unit
         if (mode & AURA_EFFECT_HANDLE_SEND_FOR_CLIENT_MASK && (apply || (!target->HasAuraType(SPELL_AURA_MOD_INCREASE_MOUNTED_FLIGHT_SPEED) && !target->HasAuraType(SPELL_AURA_FLY))))
         {
-            target->SetCanFly(apply);
-
-            if (!apply && target->GetTypeId() == TYPEID_UNIT && !target->IsLevitating())
+            if (target->SetCanFly(apply) && !apply && !target->IsLevitating()) {
```
Comment (Kitzunu) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp:3213]:
```suggestion
            if (target->SetCanFly(apply) && !apply && !target->IsLevitating()) 
            {
```

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -2877,9 +2884,9 @@ void Spell::DoAllEffectOnTarget(TargetInfo* target)
         DoTriggersOnSpellHit(spellHitTarget, mask);
 
         // if target is fallged for pvp also flag caster if a player
-        // xinef: do not flag spells with aura bind sight (no special attribute)
-        if (effectUnit->IsPvP() && effectUnit != m_caster && m_caster->GetTypeId() == TYPEID_PLAYER && !m_spellInfo->HasAura(SPELL_AURA_BIND_SIGHT))
+        if (enablePvP) {
```
Comment (Kitzunu) [inline src/server/game/Spells/Spell.cpp:2887]:
```suggestion
        if (enablePvP) 
        {
```

**Comment by felwyn** [comment]
Could someone elaborate what has to be tested? What effects should this give? I was casting a spell while mounted and it dismounted me.

**Comment by Helias** [comment]
I wrote a list of possible behaviors that this PR can break, test them all :wink:

**Comment by FrancescoBorzi** [comment]
@Helias there are conflicts

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp [hunk 3]
```diff
@@ -6328,7 +6328,10 @@ void AuraEffect::HandlePeriodicHealthLeechAuraTick(Unit* target, Unit* caster) c
     if (caster)
         caster->SendSpellNonMeleeDamageLog(target, GetId(), damage + absorb + resist, GetSpellInfo()->GetSchoolMask(), absorb, resist, false, 0, crit);
 
+<<<<<<< HEAD
 
+=======
+>>>>>>> origin/master
```
Comment (FrancescoBorzi) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp]: wrong

File: src/server/game/Spells/Spell.cpp [hunk 4]
```diff
@@ -784,7 +784,10 @@ void Spell::SelectExplicitTargets()
                 m_targets.SetUnitTarget(redirect);
                 m_spellFlags |= SPELL_FLAG_REDIRECTED;
             }
+<<<<<<< HEAD
 
+=======
+>>>>>>> origin/master
```
Comment (FrancescoBorzi) [inline src/server/game/Spells/Spell.cpp]: wrong

Comment (Helias) [reply src/server/game/Spells/Spell.cpp]: already fixed

**Comment by claudiodfc** [comment]
### Results of testing on Windows 10
[All of these tests were recorded on this video.](https://youtu.be/wmjtkJp5cGA)
No unexpected behaviour was found during this test.
- [X] use the levitate spell of the Priest and the mage one for levitating
- [X] try to fall from a high place
- [X] try to fall dismounting from a flying mount
- [X] try to fall dismounting when you are jumping from a high place and dismounting and use spells
cast spells while are you in mount
- [X] try to use spells (Melle and Ranged, so use Warrior, Hunter and Mage maybe) and use from very far (also if the client show that you can't) or from near NPCs
- [X] use Mind Control in a duel against other player and on NPC, use it once in the battleground, check if everything seems ok

**Comment by FrancescoBorzi** [comment]
now that's testing! good job @claudiodfc

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by zqgmzq** [comment]
@Helias fix plz warrior charge range check

**Comment by Helias** [comment]
I will opt to revert this asap

**Pull request merged**